Update packages (#25)
JuliaDiff/juliadiff.github.io: index.md

### Higher order

- [JuliaDiff/TaylorSeries.jl](https://github.com/JuliaDiff/TaylorSeries.jl): Taylor polynomial expansions in one or more variables
- [JuliaDiff/TaylorDiff.jl](https://github.com/JuliaDiff/TaylorDiff.jl): Higher order directional derivatives (experimental)

### Rulesets

These packages define derivatives for basic functions, and enable users to do the same:

- [JuliaDiff/ChainRules](https://www.juliadiff.org/ChainRulesCore.jl/stable/): Ecosystem for backend-agnostic forward and reverse rules.
  - [JuliaDiff/ChainRulesCore.jl](https://github.com/JuliaDiff/ChainRulesCore.jl): Core API for users to add rules to their package.
  - [JuliaDiff/ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl/): Rules for Julia Base and standard libraries.
  - [JuliaDiff/ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl/): Tools for testing rules defined with ChainRulesCore.jl.
- [ThummeTo/ForwardDiffChainRules.jl](https://github.com/ThummeTo/ForwardDiffChainRules.jl): Translate rules from ChainRulesCore.jl to make them compatible with ForwardDiff.jl

> - [JuliaDiff/DiffRules.jl](https://github.com/JuliaDiff/DiffRules.jl): Scalar rules used by ForwardDiff.jl, ReverseDiff.jl, and Tracker.jl

Reviewer: Actually, the list is much longer:
```suggestion
- [JuliaDiff/DiffRules.jl](https://github.com/JuliaDiff/DiffRules.jl): Scalar rules used by e.g. ForwardDiff.jl, ReverseDiff.jl, Zygote.jl, Tracker.jl, and Symbolics.jl

Reviewer 2: Zygote doesn't actually use DiffRules.jl except for its vestigial ForwardMode which noone uses.

Author: fixed

> [_Automatic differentiation in machine learning: a survey_](https://jmlr.org/papers/v18/17-468.html), Baydin et al. (2018)

> [_A review of automatic differentiation and its efficient implementation_](https://arxiv.org/abs/1811.05031), Margossian (2019)

## What is JuliaDiff?

JuliaDiff is an informal [GitHub organization](https://github.com/JuliaDiff/) which aims to unify and document packages written in [Julia](https://julialang.org) for evaluating derivatives.
The technical features of Julia[^1] make implementing and using differentiation techniques easier than ever before (in our biased opinion).

Discussions on JuliaDiff and its uses may be directed to the [Julia Discourse forum](https://discourse.julialang.org/).
The ChainRules project maintains a [list of recommended reading](https://www.juliadiff.org/ChainRulesCore.jl/stable/FAQ.html#Where-can-I-learn-more-about-AD-?) for those after more information.
The [autodiff.org](http://www.autodiff.org/) site serves as a portal for the academic community, though it is often out of date.

## The Big List

What follows is a big list of Julia differentiation packages and related tooling, last updated in January 2024.

> If you notice something inaccurate or outdated, please [open an issue](https://github.com/JuliaDiff/juliadiff.github.io/issues) to signal it.

Reviewer 2: Why the change from submit a PR to open an issue?

Author: Because beginners will find it easier to open an issue than to fork, clone, modify, push and PR

- [JuliaDiff/Diffractor.jl](https://github.com/JuliaDiff/Diffractor.jl): Source transformation AD backend (experimental)

### Symbolic differentiation

- [JuliaSymbolics/Symbolics.jl](https://github.com/JuliaSymbolics/Symbolics.jl): Pure Julia computer algebra system with support for fast and sparse analytical derivatives
- [brianguenter/FastDifferentiation.jl](https://github.com/brianguenter/FastDifferentiation.jl): Generate efficient executables for symbolic derivatives

### Numeric differentiation

- [JuliaDiff/FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl): Finite differences with support for arbitrary types and higher order schemes
- [JuliaDiff/FiniteDiff.jl](https://github.com/JuliaDiff/FiniteDiff.jl): Finite differences with support for caching and sparsity

### Higher order

- [JuliaDiff/TaylorSeries.jl](https://github.com/JuliaDiff/TaylorSeries.jl): Taylor polynomial expansions in one or more variables
- [JuliaDiff/TaylorDiff.jl](https://github.com/JuliaDiff/TaylorDiff.jl): Higher order directional derivatives (experimental)

> 

Reviewer 2: We should put Diffractor here too.
Diffractor's forward mode is actually a Taylor forward mode.

Author: or we could put the Taylor stuff inside forward mode? conceptually it's the same, right?

Author: I put Diffractor here

### Rulesets

These packages define derivatives for basic functions, and enable users to do the same:

- [JuliaDiff/ChainRules](https://www.juliadiff.org/ChainRulesCore.jl/stable/): Ecosystem for backend-agnostic forward and reverse rules.
  - [JuliaDiff/ChainRulesCore.jl](https://github.com/JuliaDiff/ChainRulesCore.jl): Core API for users to add rules to their package.
  - [JuliaDiff/ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl/): Rules for Julia Base and standard libraries.
  - [JuliaDiff/ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl/): Tools for testing rules defined with ChainRulesCore.jl.
- [ThummeTo/ForwardDiffChainRules.jl](https://github.com/ThummeTo/ForwardDiffChainRules.jl): Translate rules from ChainRulesCore.jl to make them compatible with ForwardDiff.jl

> - [JuliaDiff/DiffRules.jl](https://github.com/JuliaDiff/DiffRules.jl): Scalar rules used by ForwardDiff.jl

Reviewer: ```suggestion
- [JuliaDiff/DiffRules.jl](https://github.com/JuliaDiff/DiffRules.jl): Scalar rules used by ForwardDiff.jl, ReverseDiff.jl, and Tracker.jl
```

- [FluxML/ZygoteRules.jl](https://github.com/FluxML/ZygoteRules.jl): Some rules used by Zygote.jl (mostly deprecated in favor of ChainRules.jl).

### Interface

- [AbstractDifferentiation.jl](https://github.com/JuliaDiff/AbstractDifferentiation.jl): Backend-agnostic interface for algorithms that rely on derivatives, gradients, Jacobians, Hessians, etc.

### Exotic

- [JuliaDiff/SparseDiffTools.jl](https://github.com/JuliaDiff/SparseDiffTools.jl): Exploit sparsity to speed up FiniteDiff.jl and ForwardDiff.jl, as well as other algorithms.
- [gaurav-arya/StochasticAD.jl](https://github.com/gaurav-arya/StochasticAD.jl): Differentiation of functions with stochastic behavior (experimental)

### Differentiating through more stuff

Some complex algorithms are not natively differentiable, which is why derivatives have been implemented in the following packages:

- [SciML](https://github.com/SciML): For a lot of different domains of scientific machine learning: differential equations, linear and nonlinear systems, optimization problems, etc.

> - [TuringLang/DistributionsAD.jl](https://github.com/TuringLang/DistributionsAD.jl): For probability distributions

Reviewer: Not sure if this should be listed, it's a type piracy mess and the intention is to move definitions and workarounds to upstream packages such as Distributions, ReverseDiff etc.
```suggestion
```

Author: I would still mention it, if only to tell people exactly that

> - [TuringLang/DistributionsAD.jl](https://github.com/TuringLang/DistributionsAD.jl): For probability distributions

Reviewer 2: ```suggestion
- [TuringLang/DistributionsAD.jl](https://github.com/TuringLang/DistributionsAD.jl): Extensions to Distributions.jl to support AD. Somewhat obsoleted by these moving into Distributions.jl extensions.
```

Or maybe just delete it?

Author: I deleted it following @devmotion's remark
